### PR TITLE
Sscs 10138 capturing hearing channel preferences in the auto journey

### PR DESCRIPTION
**Pull request by omer-hmcts**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/SSCS-10138

### Change description ###
Only one hearing channel can be sent to HMC in individualDetail.preferredHearingChannel element, filter out options in an order of precedence based on users provided preferences

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by nickwest1 — APPROVED** [review]

**Review by nickwest1 — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 1]
```diff
@@ -140,18 +150,51 @@ public static String getIndividualLastName(Entity entity) {
         return entity.getName().getLastName();
     }
 
-    public static String getIndividualPreferredHearingChannel(String hearingType, HearingSubtype hearingSubtype) {
-        // TODO Depends on SSCS-10273 - Needs to implement for Reference data of valid Hearing Channel codes
-        return null;
+    public static Optional<String> getIndividualPreferredHearingChannel(String hearingType,
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:153]: Can you change the return to a Sring

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 2]
```diff
@@ -140,18 +150,51 @@ public static String getIndividualLastName(Entity entity) {
...
+    private static boolean shouldPreferTelephoneHearingChannel(HearingSubtype hearingSubtype) {
+        return isYes(hearingSubtype.getWantsHearingTypeTelephone()) && nonNull(hearingSubtype.getHearingTelephoneNumber());
+    }
+
+    private static boolean shouldPreferVideoHearingChannel(HearingSubtype hearingSubtype) {
+        return isYes(hearingSubtype.getWantsHearingTypeVideo())
+            && nonNull(hearingSubtype.getHearingVideoEmail());
+    }
+
+    public static Optional<String> getIndividualInterpreterLanguage(HearingOptions hearingOptions) {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:180]: Same here, can you return a String or null

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 3]
```diff
@@ -140,18 +150,51 @@ public static String getIndividualLastName(Entity entity) {
...
+                .map(SignLanguage::getHmcReference);
+        }
+        if (isYes(hearingOptions.getLanguageInterpreter())) {
+            return getInterpreterLanguage(hearingOptions)
+                .map(InterpreterLanguage::getHmcReference);
+        }
+        return Optional.empty();
+    }
+
+    private static Optional<SignLanguage> getSignLanguage(HearingOptions hearingOptions) {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:192]: Same here, can you return a String and add orElse(null) at the end.

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 4]
```diff
@@ -140,18 +150,51 @@ public static String getIndividualLastName(Entity entity) {
...
+        return Optional.ofNullable(SignLanguage.getSignLanguageKeyByCcdReference(hearingOptions.getSignLanguageType()));
+    }
+
+    private static Optional<InterpreterLanguage> getInterpreterLanguage(HearingOptions hearingOptions) {
+        return Optional.ofNullable(InterpreterLanguage.getLanguageAndConvert(hearingOptions.getLanguages()));
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:197]: Same here, can you return a String and add orElse(null) at the end.

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 5]
```diff
@@ -140,18 +150,51 @@ public static String getIndividualLastName(Entity entity) {
...
+
+    private static boolean shouldPreferVideoHearingChannel(HearingSubtype hearingSubtype) {
+        return isYes(hearingSubtype.getWantsHearingTypeVideo())
+            && nonNull(hearingSubtype.getHearingVideoEmail());
+    }
+
+    public static Optional<String> getIndividualInterpreterLanguage(HearingOptions hearingOptions) {
+        if (isTrue(hearingOptions.wantsSignLanguageInterpreter())) {
+            return getSignLanguage(hearingOptions)
+                .map(SignLanguage::getHmcReference);
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:183]: if getSignLanguage returns null or an invalid string, should it throw an error or continue to check to see if getInterpreterLanguage is valid before returning null?

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 1 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java]: Changed return type to String, default return is now to throw IllegalStateException as discussed over teams, we can expect only valid input to be sent in from SYA  as the prior flow would not proceed without it.

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 2 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:188]: Done

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 3 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java]: Done

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 4 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java]: Done

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 6]
```diff
@@ -64,15 +64,33 @@ void buildHearingDetails() {
         assertNotNull(hearingDetails.getPanelRequirements());
     }
 
-    @DisplayName("shouldBeAutoListed Parameterized Tests")
+    @DisplayName("shouldBeAutoListed flag is TRUE when HearingOptions.wantsToAttend is yes")
     @Test
     void shouldBeAutoListed() {
         // TODO Finish Test when method done
```
Comment (johnoneill46) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:70]: Can this be removed?

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java [hunk 7]
```diff
@@ -58,9 +58,11 @@ public static HearingDetails buildHearingDetails(HearingWrapper wrapper) {
         return hearingDetailsBuilder.build();
     }
 
-    public static boolean shouldBeAutoListed() {
-        // TODO Future Work
-        return true;
+    public static boolean shouldBeAutoListed(SscsCaseData caseData) {
+        // TODO Future Work - Account for additional conditions
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java:62]: I'd strip this comment

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 8]
```diff
@@ -150,36 +150,56 @@ public static String getIndividualLastName(Entity entity) {
         return entity.getName().getLastName();
     }
 
-    public static Optional<String> getIndividualPreferredHearingChannel(String hearingType, HearingSubtype hearingSubtype) {
+    public static String getIndividualPreferredHearingChannel(String hearingType,
+                                                                        HearingSubtype hearingSubtype,
+                                                                        HearingOptions hearingOptions) {
         if (hearingType == null || hearingSubtype == null) {
-            return Optional.empty();
+            return null;
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:157]: This should probably be an IllegalStateException too

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 9]
```diff
@@ -496,57 +507,93 @@ void getIndividualInterpreterSignLanguageTest(String signLang, String expected)
...
+        HearingSubtype hearingSubtype = HearingSubtype.builder().wantsHearingTypeTelephone("Yes").build();
+        HearingOptions hearingOptions = HearingOptions.builder().wantsToAttend("yes").build();
+
+        IllegalStateException thrown = assertThrows(
+            IllegalStateException.class,
+            () -> getIndividualPreferredHearingChannel("oral", hearingSubtype, hearingOptions),
+            "Expected doThing() to throw, but it didn't"
+        );
+
+        assertTrue(thrown.getMessage().contains("Failed to determine a preferred hearing channel"));
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:561]: (Minor) can use assertJ's assertThatExceptionOfType(IllegalStateException.class).isThrownBy(()-> getIndividualPreferredHearingChannel) to do the assertion in on eline

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java [hunk 10]
```diff
@@ -496,57 +507,93 @@ void getIndividualInterpreterSignLanguageTest(String signLang, String expected)
...
         assertThat(result).isEqualTo(HearingChannel.FACE_TO_FACE.getHmcReference());
     }
 
+    @DisplayName("When wantsToAttend is yes, and wantsHearingType video but hearingVideoEmail is not set throw IllegalStateException")
+    @Test
+    void getIndividualPreferredHearingChannelNullWhenMissingPartialRequirementsVideoExample() {
+        HearingSubtype hearingSubtype = HearingSubtype.builder().wantsHearingTypeVideo("Yes").build();
+        HearingOptions hearingOptions = HearingOptions.builder().wantsToAttend("yes").build();
+
+        IllegalStateException thrown = assertThrows(
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java:588]: Same again

**Review by dattipoeHMCTS — CHANGES_REQUESTED** [review]

**Review by dattipoeHMCTS — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java (hunk 6 above)
Comment (omer-hmcts) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java]: Done

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 8 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java]: Done

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java (hunk 7 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java]: Done

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java (hunk 9 above)
Comment (omer-hmcts) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java]: Done

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java (hunk 10 above)
Comment (omer-hmcts) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMappingTest.java]: Done

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 11]
```diff
@@ -175,17 +175,42 @@ public static String getIndividualFullName(Entity entity) {
         return entity.getName().getFullNameNoTitle();
     }
 
+    public static String getIndividualPreferredHearingChannel(String hearingType,
+                                                                        HearingSubtype hearingSubtype,
+                                                                        HearingOptions hearingOptions) {
+        if (eitherNull(hearingType, hearingSubtype)) {
+            throw new IllegalStateException("hearingType and/or hearingSubtype null");
+        }
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:183]: This shouldn't be here.
HearingType will be null for OtherParty. 
HearingType and hearingSubtype will both be null for JointParty

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 12]
```diff
@@ -82,7 +82,7 @@ public static List<PartyDetails> buildHearingPartiesDetails(SscsCaseData caseDat
             for (CcdValue<OtherParty> ccdOtherParty : otherParties) {
                 OtherParty otherParty = ccdOtherParty.getValue();
                 partiesDetails.addAll(buildHearingPartiesPartyDetails(
-                        otherParty, otherParty.getRep(), otherParty.getHearingOptions(), null, otherParty.getHearingSubtype(), appellant.getId()));
+                        otherParty, otherParty.getRep(), otherParty.getHearingOptions(), appeal.getHearingType(), otherParty.getHearingSubtype(), appellant.getId()));
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:85]: otherParty shouldn't be using appeal.getHearingType(), that is for the appellant.

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 11 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:187]: Being addressed here: **https://github.com/hmcts/sscs-hearings-api/pull/185**

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java (hunk 12 above)
Comment (omer-hmcts) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java]: Being addressed here: https://github.com/hmcts/sscs-hearings-api/pull/185

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java [hunk 13]
```diff
@@ -180,16 +179,42 @@ public static String getIndividualFullName(Entity entity) {
         return entity.getName().getFullNameNoTitle();
     }
 
-    public static Optional<String> getIndividualPreferredHearingChannel(String hearingType, HearingSubtype hearingSubtype) {
-        if (hearingType == null || hearingSubtype == null) {
-            return Optional.empty();
+    public static String getIndividualPreferredHearingChannel(String hearingType,
+                                                                        HearingSubtype hearingSubtype,
+                                                                        HearingOptions hearingOptions) {
+        if (eitherNull(hearingType, hearingSubtype)) {
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsPartiesMapping.java:185]: Don't think you need this as a whole method. Re: your comment below Harry's fix has gone in so you'd be overwriting it

**Review by dattipoeHMCTS — CHANGES_REQUESTED** [review]

**Review by dattipoeHMCTS — APPROVED** [review]

**Review by nickwest1 — APPROVED** [review]

**Pull request merged**